### PR TITLE
proposal for admission webhook

### PR DESCRIPTION
**Pull request by chendave**
Signed-off-by: Dave Chen <dave.chen@arm.com>

/kind design
/kind documentation
/kind feature

### CODE REVIEW

**Comment by chendave** [comment]
cc @fisherxu  @rohitsardesai83 @sids-b

File: docs/proposals/web-hook.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+			CABundle: context.signingCert,
+		},
+	},
+	...
+	}
+```
+
+
+## Certification management
+Need to generate another set of key / certification, or reuse the existing key / certification for API validation.
```
Comment (rohitsardesai83) [inline docs/proposals/web-hook.md]: Can we refer from here : https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers

File: docs/proposals/web-hook.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+As the evolving of the project, it is foreseeable more API and resouce will be added in the project. kubeedge so far lack of effective way of pre-processing for the object configuration, for example, whether the pod that is going to be created contains the unwanted label, could be the specific configmap be protected from deletion etc.
+
+There is a concrete example on the github [issue 845](https://github.com/kubeedge/kubeedge/issues/845), the issue there cannot be addressed by [CRD validation](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation), so we need to explore approach to achieve that purpose.
+
+
+## Goals
+* A framework to enable admission control in kubeedge.
+* Address the edge cases where kubernetes CRD validation cannot accomplish.
+* Create basic integration testcases.
+
```
Comment (rohitsardesai83) [inline docs/proposals/web-hook.md]: Can we specify some non-goals here ? What problems are we not trying to solve with this approach ?

File: docs/proposals/web-hook.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+Propose using Kubernetes [Dynamic Admission Control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) to determine whether to accept or deny the request, the decision is based on how is the policy is configured, this give us chance to validate the request before persisting the object.
+
+
+# Design Details
+
+## Admission controller on control-plane
+
+Create a new admission control when the module `controller` is started.
+```go
+func (ctl *Controller) Start(c *bcontext.Context) {
```
Comment (fisherxu) [inline docs/proposals/web-hook.md]: The admission controller is not associated with cloud part(cloudhub/controller), it should be started independently.
Where the code base placed， let's disscuss it, how do you think? @chendave @kevin-wangzefeng @rohitsardesai83 @sids-b

Comment (chendave) [reply docs/proposals/web-hook.md]: Do you mean we need another module beside "cloudhub", "controller", and "devicecontroller"?

Comment (chendave) [reply docs/proposals/web-hook.md]: is that too heavy?

**Comment by sids-b** [comment]
Thanks @chendave 

/approve

**Comment by sids-b** [comment]
Please change the status to implementable from Alpha

**Comment by chendave** [comment]
@sids-b , done, the design has been drifted away from the implementation, shall we update the content with current implementation?

**Comment by chendave** [comment]
@sids-b  @fisherxu  updated, could you pls take a another look?

**Comment by fisherxu** [comment]
/lgtm
Thanks!  @chendave

**Pull request merged**